### ISSUE TITLE
Exception raising `Stdlib.at_exit` callback breaks `Stdlib.exit`

### ISSUE THREAD

**Issue by kayceesrk** [ocaml/ocaml#11221]
Consider the following program:

```ocaml
Stdlib.at_exit (fun _ -> failwith "exn");;
try
  Stdlib.exit 0
with _ -> assert false (* reaches here *)
```

The callback registered with `at_exit` raises an exception. Since `Stdlib.exit` runs the `at_exit` callbacks, the program does not get to exit the program and the control goes to the exception handler. The program fails with

```bash
Fatal error: exception Assert_failure("exit_fail.ml", 4, 10)
```

Is this reasonable behaviour? What I find unconvincing is that there is no way to exit the program now using `Stdlib.exit` as there is no way to circumvent the `at_exit` callbacks. 

This issue exists in OCaml 4 as well.

**Comment by dbuenzli** [comment]
Related to https://github.com/ocaml/ocaml/issues/7253

> Is this reasonable behaviour?

I think a first good step would be to trap them with the uncaught exception trap. AFAIR I tried at some point in https://github.com/ocaml/ocaml/pull/685 but got into recursive dependency problems – this may have changed.

More broadly a better semantics for `exit` handlers raising or `exit`ing would be [this](https://github.com/ocaml/ocaml/pull/685#issuecomment-337956714).

**Comment by xavierleroy** [comment]
Right, this was discussed  at length in #7253.  As far as I can remember, the consensus (?) at the time was that an exception escaping out of an `at_exit` handler is a programming mistake, but must not be suppressed silently.  Instead, it should escape the `Stdlib.exit` so as to be treated by the handler for uncaught exceptions.  The latter will call `Stdlib.exit` again, but we make sure the `at_exit` handler is not raised again.

In conjunction with domains, this doesn't work very well if a domain other than the initial domain calls `Stdlib.exit`.  Not sure what we should do about that.  But let's try not to repeat old discussions if at all possible.

**Comment by dbuenzli** [comment]
> Instead, it should escape the `Stdlib.exit` so as to be treated by the handler for uncaught exceptions.

That's not what I took from the discussions. I think people generally agreed that it was better if `Stdlib.exit` acted as a blackhole `int -> 'a` function (e.g. @alainfrisch [here](https://github.com/ocaml/ocaml/pull/685#issuecomment-232177702))  and that it was in fact rather surprising that you could "come back" from it.

Propagating exceptions occurring in `at_exit` functions in the context where `Stdlib.exit` is being called does not make any sense in my opinion. Rather, the code in charge of executing the `at_exit` handlers should simply catch and trap them to the uncaught exception handler directly.

**Comment by kayceesrk** [comment]
`at_exit` callbacks throwing exceptions smells like a programming error. Trapping such exceptions at the uncaught exception handler feels cleanest. Such a solution may be folded into the uncaught exception handlers being discussed at https://github.com/ocaml/ocaml/issues/11074.

**Comment by gadmm** [comment]
Such a generalization of the uncaught exception handler was proposed at https://github.com/ocaml/ocaml/pull/8962 for similar purposes (exception raised during clean-up denoting a programming error).
* There was the question of what to do in this uncaught handler. The user might want to continue, with or without logging the problem, or stop the program (it is not possible to handler the exception "programmatically"). I made it customizable in this way. In Rust/C++, the program is aborted; the termination handler is customizable but it cannot return to its caller.
* You want to turn off async exceptions during the uncaught exception handler, and the at_exit handlers, because those are not programming errors if they occur during clean-up (and it does not make sense to interrupt resource clean-up). So you need something like #8961.
